### DISCUSSION TITLE
[General] gRPC client interface

### DISCUSSION THREAD

**Discussion by jpohanka** [trinodb/trino#13727] [General]
**Issue type**: feature

**Trino version**: 363

**Description**:

Currently, Trino supports two client interfaces:
- JDBC driver
- REST API

It would be benefitial to support a **gRPC client interface** due to its speed advantage over REST API and its cross-platform advantage over JDBC.

**Comment by martint** [comment]
Why would gRPC be faster than the HTTP-based API?

**Comment by hashhar** [comment]
The transport being changed won't help unless we also change the protocol so that multiple parallel fetches can be made to the output buffer on the coordinator.

Also gRPC isn't very good to work with in all languages and makes writing clients harder.

**Reply by nineinchnick** [reply]
gRPC is basically protobuf over http/2. We don't need the RPC part, so I'd reduce the discussion to replacing json with another format. But as @hashhar just said, there's more design needed than just that.

**Reply by hashhar** [reply]
+1 Jan.

Changing formats may be a quick win but it adds work to update all clients to work with it. And having multiple such breakages (format change, then transport change, then some protocol change) seems wasteful than going to the step which would win us most performance at once.

Note that there are community contributed clients as well over which we have no direct control so we'd want to make sure we are conservative about the work we make for others.

**Comment by martint** [comment]
See https://github.com/trinodb/trino/issues/3042

**Comment by jpohanka** [comment]
@martint @hashhar @nineinchnick Thank you for the feedback.

**Comment by hackeryang** [comment]
Maybe we can try `Thrift` RPC(we have drift in airlift framework) in a long term future?  

It seems that the performance of thrift communication is better than HTTP, but it will use more native memory: https://github.com/prestodb/presto/pull/13894
